proxy/dialer_test.go

```diff
@@ -0,0 +1,133 @@
+package core
+
+import (
+	"testing"
+	"net/http"
+	"golang.org/x/net/proxy"
+	"github.com/shadowsocks/go-shadowsocks2/socks"
+	"github.com/shadowsocks/go-shadowsocks2/core"
+	"io/ioutil"
+	"net"
+	"io"
+	"time"
+	"strings"
+)
+
+func TestDialerInterface(t *testing.T) {
+	var dialer proxy.Dialer
+	dialer, err := NewDialer("198.51.100.254:65534", "aes-128-gcm", "dialer-test")
+	if err != nil {
+		t.Fatal("Unable to create new dialer:", err.Error())
+		return
+	}
+	// AES-128 key is 16 bytes long.
+	// This key is just for interface testing only, so zeroKey doesn't matter.
+	zeroKey := make([]byte, 16)
+	dialer, err = NewDialerFromKey("195.51.100.254:65533", "aes-128-gcm", zeroKey)
+	if err != nil {
+		t.Fatal("Unable to create new dialer with pre-generated key:", err.Error())
+		return
+	}
+	_ = dialer
+	t.Log("Dialer interface works.")
+}
+
+func TestDialerFunctionality(t *testing.T) {
+	// First let's launch a dummy shadowsocks server.
+	ciph, err := core.PickCipher("aes-128-gcm", nil, "dialer-test")
+	if err != nil {
+		t.Fatal("Unable to pick cipher", ciph, ":", err.Error())
+		return
+	}
+	go tcpRemote("127.0.0.1:60001", ciph.StreamConn)
+	dialer, err := NewDialer("127.0.0.1:60001", "aes-128-gcm", "dialer-test")
+	if err != nil {
+		t.Fatal("Unable to create new dialer:", err.Error())
+		return
+	}
+	client := &http.Client {
+		Transport: &http.Transport {
+			Dial: dialer.Dial,
+		},
+	}
+	resp, err := client.Get("http://ifconfig.co")
```

Comment:
Your unit tests shouldn't depend on external services.

Why don't you run a simple TCP service locally instead?
Here is an example where I do that: https://github.com/Jigsaw-Code/outline-ss-server/blob/master/shadowsocks/tcp_test.go